scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -291,8 +294,23 @@ object Execution {
      */
     def finished(): Unit = messageQueue.put(Stop)
 
-    def getOrElseInsert[T](ex: Execution[T],
-      res: => Future[(T, ExecutionCounters)])(implicit ec: ConcurrentExecutionContext): Future[(T, ExecutionCounters)] = {
+    def getOrLock(write: ToWrite): Either[Promise[ExecutionCounters], Future[ExecutionCounters]] = {
+      /*
+       * Since we don't want to evaluate res twice, we make a promise
+       * which we will use if it has not already been evaluated
+       */
+      val promise = Promise[ExecutionCounters]()
+      val fut = promise.future
+      toWriteCache.putIfAbsent(write, fut) match {
```

Comment:
do we want to make our two subclasses of ToWrite case classes since they are used as hashkeys now?

```
case class SimpleWrite[T](pipe: TypedPipe[T], sink: TypedSink[T]) extends ToWrite {
 ...
}

case class PreparedWrite[T](fn: (Config, Mode) => SimpleWrite[T]) extends ToWrite {
 ...
}
```